src/operator/tensor/indexing_op.h

```diff
@@ -0,0 +1,169 @@
+                                     << (*in_type)[i];
+    }
+  }
+  out_type->clear();
+  out_type->push_back(dtype);
+  return true;
+}
+
+template<typename xpu>
+void EmbeddingOpForward(const nnvm::NodeAttrs& attrs,
+                        const OpContext& ctx,
+                        const std::vector<TBlob>& inputs,
+                        const std::vector<OpReqType>& req,
+                        const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mshadow::expr;
+  CHECK_EQ(req[embedding::kOut], kWriteTo);
+  CHECK_EQ(inputs.size(), 2);
+  CHECK_EQ(outputs.size(), 1);
+  CHECK_EQ(inputs[embedding::kWeight].ndim(), 2)
+          << "Embedding layer expects its weight to be two-dimensional. "
+          << inputs[embedding::kWeight].ndim()
+          << " dimensional input is given instead";
+
+  const TShape& ishape = inputs[embedding::kData].shape_;
+  const TShape& oshape = outputs[embedding::kOut].shape_;
+
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+    Tensor<xpu, 1, DType> data = inputs[embedding::kData].get_with_shape<xpu, 1, DType>(
+      Shape1(ishape.ProdShape(0, ishape.ndim())), s);
+    Tensor<xpu, 2, DType> wmat = inputs[embedding::kWeight].get<xpu, 2, DType>(s);
+    Tensor<xpu, 2, DType> out = outputs[embedding::kOut].get_with_shape<xpu, 2, DType>(
+      Shape2(oshape.ProdShape(0, oshape.ndim()-1), oshape[oshape.ndim()-1]), s);
+    out = take(data, wmat);
+  });
+}
+
+template<typename xpu>
+void EmbeddingOpBackward(const nnvm::NodeAttrs& attrs,
+                         const OpContext& ctx,
+                         const std::vector<TBlob>& inputs,
+                         const std::vector<OpReqType>& req,
+                         const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mshadow::expr;
+  CHECK_EQ(inputs.size(), 3);
+  CHECK_EQ(outputs.size(), 2);
+  CHECK_EQ(req[embedding::kData], kNullOp)
+          << "Embedding layer doesn't support calculate data gradient";
+
+  const TShape& ishape = inputs[1].shape_;
+  const TShape& oshape = inputs[0].shape_;
+
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+    Tensor < xpu, 1, DType > data = inputs[1].get_with_shape<xpu, 1, DType>(
+      Shape1(ishape.ProdShape(0, ishape.ndim())), s);
+    Tensor<xpu, 2, DType> grad_out = inputs[0].get_with_shape<xpu, 2, DType>(
+    Shape2(oshape.ProdShape(0, oshape.ndim()-1), oshape[oshape.ndim()-1]), s);
+    Tensor<xpu, 2, DType> grad_in = outputs[1].get<xpu, 2, DType>(s);
+
+
+    if (req[embedding::kWeight] == kWriteTo || req[embedding::kWeight] == kAddTo) {
+      if (req[embedding::kWeight] == kWriteTo) {
+#ifdef __CUDACC__
+        cudaMemsetAsync(grad_in.dptr_, 0, grad_in.MSize() * sizeof(DType),
+                          Stream<gpu>::GetStream(s));
+#else
+        grad_in = scalar<DType>(0.0f);
+#endif
```

Comment:
Seems that the nested macro here causes some compilation problem on windows. We could delete the `cudaMemsetAsync` part and retry.